cli_tools/common/utils/logging/firelog_logger.go

```diff
@@ -0,0 +1,311 @@
+func RunWithServerLogging(action string, params InputParams, function func() (*daisy.Workflow, error)) {
+	l := NewFirelogLogger(action, params)
+	l.runWithServerLogging(function)
+}
+
+func (l *FirelogLogger) sendLogToServer(logEvent *ComputeImageToolsLogExtension) logResult {
+	return l.sendLogToServerWithRetry(logEvent, 3)
+}
+
+func (l *FirelogLogger) sendLogToServerWithRetry(logEvent *ComputeImageToolsLogExtension, maxRetry int) logResult {
+	logMutex.Lock()
+	defer logMutex.Unlock()
+
+	for i := 0; i < maxRetry; i++ {
+		// Before sending a new request, wait for a while if server asked to do so
+		if nextRequestWaitMillis > 0 {
+			nextRequestWaitMillis = 0
+			time.Sleep(time.Duration(nextRequestWaitMillis) * time.Millisecond)
+		}
+
+		logRequestJSON, err := constructLogRequest(logEvent)
+		fmt.Println(string(logRequestJSON))
+		if err != nil {
+			fmt.Println("Failed to log to server: failed to prepare json log data.")
+			return failedOnCreateRequestJSON
+		}
+
+		if !serverLogEnabled {
+			return serverLogDisabled
+		}
+
+		req, err := http.NewRequest("POST", l.ServerURL, bytes.NewBuffer(logRequestJSON))
+		if err != nil {
+			fmt.Println("Failed to create log request: ", err)
+			return failedOnCreateRequest
+		}
+
+		req.Header.Set("Content-type", "application/json")
+		req.Header.Set("X-Goog-Api-Key", reverse(key))
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			fmt.Println("Failed to log to server: ", err)
+			continue
+		}
+
+		defer resp.Body.Close()
+		body, _ := ioutil.ReadAll(resp.Body)
+		var lr LogResponse
+		if err = json.Unmarshal(body, &lr); err != nil {
+			fmt.Println("Failed to parse log response: ", err, "\nResponse: ", string(body))
+			return failedToParseResponse
+		}
+
+		// Honor "NextRequestWaitMillis" from server for traffic control. However, wait no more than 5s to prevent a long
+		// stuck
+		if lr.NextRequestWaitMillis > 0 {
+			nextRequestWaitMillis = lr.NextRequestWaitMillis
+			if nextRequestWaitMillis > 5000 {
+				nextRequestWaitMillis = 5000
+			}
+		}
+
+		// Honor "ResponseAction" from server to control retrying
+		if len(lr.LogResponseDetails) > 0 {
+			action := lr.LogResponseDetails[0].ResponseAction
+			if action == DeleteRequest || action == ResponseActionUnknown {
+				// Log success or unknown status, just return
+				return logResult(action)
+			} else if action == RetryRequestLater {
+				// Retry as server asked
+				continue
+			}
+			// Return if client failed to receive a defined response action
+			fmt.Println("Failed to log to server: undefined response action: ", action)
+			return failedOnUndefinedResponse
+		}
+
+		// Return if client failed to receive response details from server
+		fmt.Println("Failed to log to server: missing response details")
+		return failedOnMissingResponseDetails
+	}
+
+	fmt.Println("Failed to log to server after retrying")
+	return failedAfterRetry
+}
+
+// Hash a given string for obfuscation
+func Hash(s string) string {
```

Comment:
extract to a method in string_utils

Reply:
The Hash result is impacted by the key, so I don't think it should be shared across other usage. We shall use it for only firelog obfuscating.

Reply:
The hash key can be a parameter in this method. You can also have a utility in logger that would populate this parameter so it's not duplicated, but hashing could still be used in other utils

Reply:
Isn't that 'other utils' just 'highwayhash' lib? I don't see an obvious benefit to wrap it as a shared utils of our repository...

Reply:
It might or might not be needed in other wrappers. I would extract anything that is not specifically related to the package where code resides and there's a reasonable chance it would be reused by another piece of code. This is a best practice IMO

Reply:
I totally understand the thinking, it's generally the best practice. Here it's special because: usually if we want to wrap it as a shared utils, we expect to abstract the interface and hide the details so that we may change the implementation without impacting callers. However, here we don't want to hide what is the concrete hash lib. Each hash algorithm, for example, "highwayhash", can lead to different hash results. We expect a stable hash result in our log so that we can use it for grouping. While it didn't hide any other details (it's just 2 lines of code) but just hide the hash algorithm, I don't consider it as a proper utils.
Let me know whether that sounds correct.